### DISCUSSION TITLE
Ideas and support for those new to Earth Data Science and coding

### DISCUSSION THREAD

**Discussion by brglea** [earthlab-education/Earth-Analytics-AY24#127] [General]
I am a newbie to earth data science and coding in general and wondering what others think helped them when first entering this space? Did you read a lot of articles, did you ask a lot of questions etc. Any and all thoughts and feels welcome!

I feel like a fish out of water and am seeing others in this course have experience or already work in this field or something adjacent. If you were to start from zero, or close to zero, what suggestions would you have?

**Comment by elyna-grapstein** [comment]
I'm in the same boat! Not much to add here other than comradery.

**Comment by akpost21** [comment]
@brglea and @elyna-grapstein welcome to earth data science and coding! I'm an ecologist that has only learned coding and GitHub in the last few years, so I know the growing pains... It can definitely be overwhelming at first, but I'm confident you'll get it. Just keep trying things- I learned the most when things didn't work and I had to understand why. I definitely learn the most from doing. Also, please feel free to always ask questions! Happy to have you in the course :-)

**Reply by akpost21** [reply]
And ChatGPT can be a great resource when you get really stuck!

**Comment by allenmoench** [comment]
I'm also in the same boat! Still trying to work out what a pull request is (I understand it conceptually, but not sure how to submit one). Also trying to figure out why github is sending me a ton of emails, most of which are irrelevant (I may have fixed that with the notification settings already), and other things. I'm sure I'll get used to the structure of this thing before too long, but in the meantime it's somewhat confusing. Anyone know how to submit a pull request? I'm trying to pull in the changes from the main branch to my own branch, to update it so that it's current.

**Reply by nquarder** [reply]
Hi @allenmoench - Try playing around in your settings to see what you can change to avoid getting so many emails! ;) 

Also, re: Pull Requests

We did examples of those in the Section 1 and 2 Discussion sections last week. Have a look at those recordings and see if you can follow the examples. Also, there are materials [in the textbook](https://cu-esiil-edu.github.io/esiil-learning-portal/foundations/pages/03-git-github/02-github-collaboration/06-pr-activity-fork.html) that demonstrate how to do a Pull Request. If you complete this week's assignments Hometowns/First-Map you will also get some more practice. 

Great questions and keep up the good work! -NQ

**Reply by byandell** [reply]
@allenmoench, On the repo page <https://github.com/earthlab-education/Earth-Analytics-AY24>  to the right of the repo name (below the horizontal task list) click the box marked "Watch". You can then change to "Participating and @mentions". That should reduce substantially. If you change to "Ignore", you will get no emails, but that might mean you miss key notifications from instructors.

**Reply by allenmoench** [reply]
Great, thanks Nathan and Brian! I found the section in the textbook that describes pull requests, so should be able to figure it out.

**Comment by nquarder** [comment]
Hi @brglea and great question here. We were all new at one point and we're all still learning. It is totally normal to feel like a fish out of water with this material, and you are in the right place. Based on my quick survey it looks we have quiet a variety of backgrounds and experiences with coding and Earth Data Science in this course.

I'll point to a few resources that you'll want to bookmark as we get rolling: 
- [Course Textbook](https://cu-esiil-edu.github.io/esiil-learning-portal/foundations/pages/00-course-overviews/foundations/00-home.html)  
- [Earth Data Science Textbook ](https://www.earthdatascience.org/)

When I was first learning R, Python, GitHub and Markdown I did a lot of Googling around to see if others had asked similar questions to those that I was trying to answer. I spent a lot of time looking through discussion boards in places like [Stackoverflow](https://stackoverflow.com/) and [Stackexchange](https://stackexchange.com/) to see if I could find someone else's code, copy it, and then make modifications based on what I was trying to accomplish. Honestly, I didn't write a lot of code from scratch (and still don't!) but, rather, take code that others have written, or that I was working from previously, and then make changes. This is how we'll get you started with this week's assignments!

And, like @akpost21 says, now, with tools like [ChatGPT](https://chatgpt.com/), I find myself using it a lot to help me write code or get started on a difficult task. For those unfamiliar with ChatGPT, it is what is called a Large Language Model (LLM), that is trained on vast amounts of internet text to help answer questions and make working more efficient. It's kind of like Google but can be used for more advanced help sessions. We will demonstrate how to use these tools throughout the course.  

For me, personally, I didn't read a lot of articles, but did ask a lot of questions, and attended a lot of great workshops and professional development. So, again, @brglea, I think you're in the right place and are off to a great start by creating this discussion. Keep it up and we look forward to working with you! -NQ

**Comment by eculler** [comment]
@brglea great question, and I'll echo that lots of folks are in the same spot right now! We have a process that works pretty well -- and @nquarder has the data to prove it! But we also know that it's tough the first few weeks or months. A couple things to bear in mind:

- You can think about learning to code as like going to a gym. You're not going to run a marathon or whatever your goal is the first time you try -- and even a little bit can feel terrible. But if you keep running a little every day it starts to feel better and soon enough you're doing things you didn't think were possible. So we're asking you to trust the process here, just like you would with sports or music or any other new skill.
- We use project-based learning in this program. We think it's more fun and more meaningful, but also the research is unequivocal that it's necessary in learning to code. In one experiment, students were split into two groups: one doing hands-on coding and the other attending lectures. The students doing projects were less happy, and felt they were learning less...and they blew the lecture group out of the water on the exam. We've seen enough brand new coders in the classroom to know you'll blow us away by the end of the semester, even if it doesn't feel like that right now (or ever)
- That brings me to my third point. I know this isn't exactly what you asked, but I never miss an opportunity to say this to the class because it's important! **Things are going to go wrong when you code.** Things go wrong when I code, and when Nate and Alison and Lilly code, and when anyone of any skill level codes. The skill of coding is not knowing everything about coding and getting it right the first time -- it's being able to calmly break down the problems that will _definitely_ happen, and find a solution. If we were able to arrange it so that you never ran into a bug or anything unexpected, we wouldn't be teaching you to code at all. 

So there's my soapbox! You'll definitely hear me say all that again throughout the semester. 

I personally have always learned to code by trying to do stuff and searching the internet or ChatGPT for examples, but we do hear from lots of students who find articles and books to be helpful. In addition to our websites that @nquarder linked to, which are ✨awesome✨ 😜 we've heard good things about the coding books from O'Reilly. One that I like is [Think Python](https://www.greenteapress.com/thinkpython/thinkpython.html) by one of my professors back in the day Allen Downey, which is free to download. Fun fact -- Allen actually wrote the book Think Java (Java is another programming language), and then a reader translated it into Python and he learned Python from his own book! I also like to read articles on Medium and Toward Data Science, although the quality varies there. And my last suggestion is a book about debugging written as a series of mysteries called [_If I Only Changed the Software, Why is the Phone on Fire?_](https://www.goodreads.com/book/show/1041825.If_I_Only_Changed_the_Software_Why_is_the_Phone_on_Fire_) by Lisa Simone. We're not working with embedded code here, but it still does a great job of illustrating how to think like a coder and solve problems.

So there's some resources I like. I'm sure others will have different ideas. In the meantime, stick with it -- nothing can substitute for experience and trying to complete projects. And we're always available to help if you get stuck!

**Comment by lauren-pease** [comment]
For someone new to coding, I would recommend going through the getting started tutorial associated with a language or library. In addition, if you have a question regarding how to do something or around syntax, I would check [StackOverflow](https://stackoverflow.com) and reference documentation.

The [pandas](https://pandas.pydata.org/docs/index.html) library for instance: 

- [Tutorial](https://pandas.pydata.org/docs/user_guide/10min.html#min)
- [Reference documentation](https://pandas.pydata.org/docs/reference/index.html)

**Reply by nquarder** [reply]
Great suggestions @lauren-alexandra! I'd also give a plug for our lessons on pandas in our old EDS textbook: 
https://www.earthdatascience.org/courses/intro-to-earth-data-science/scientific-data-structures-python/pandas-dataframes/

This will look similar to the steps you're doing for this week's assignment. -NQ

**Comment by BioNCphilic** [comment]
@brglea - I'm curious.... how are you feeling now? 
I think playing around with the code and testing things out is helpful. Just looking at code helps focus your attention on a , or a (), but the real learning comes by understanding what happens when you run the code and error messages can be a great learning tool once you get comfortable with the fact that they are a guide not an assessment of what you are trying to do. 

In any case, I hope you're feeling more comfortable. Happy coding!
